### PR TITLE
Enable Qwen3 and Deepseekv3 mxfp8 Training

### PR DESCRIPTION
**Pull request by sudhu2k**
# What does this PR do?
Adds training scripts for Qwen3 and modifies existing Deepseekv3 script to enable mxfp8 training.

### CODE REVIEW

File: examples/deepseek_v3/train_deepseekv3.sh [hunk 1]
```diff
@@ -133,13 +135,24 @@ fi
...
+fi
 echo ""
 
 # perf options
 OPTIMIZER_OFFLOAD=false
 GEMM_TUNING="${GEMM_TUNING:-1}"
 USE_GROUPED_GEMM="${USE_GROUPED_GEMM:-true}"
 MOE_USE_LEGACY_GROUPED_GEMM="${MOE_USE_LEGACY_GROUPED_GEMM:-true}"
+# FP8 MoE requires TE grouped GEMM; legacy grouped_gemm does not implement FP8 expert matmuls.
+if [ "$PR" = fp8 ]; then
```
Comment (wenchenvincent) [inline examples/deepseek_v3/train_deepseekv3.sh:150]: Which group GEMM do we use by default?
Seems fp8 and mxfp8 uses the multi-stream hipblaslt from TE.

For bf16, are we using the legacy grouped gemm by default? Shall we use the bf16 CK group gemm in TE?

File: examples/deepseek_v3/train_deepseekv3.sh [hunk 2]
```diff
@@ -164,15 +177,25 @@ PROFILE_SYNC=${PROFILE_SYNC:-false}
 PROFILE_START=${PROFILE_START:-6}
 PROFILE_END=${PROFILE_END:-7}
 FORCE_BALANCE=${FORCE_BALANCE:-false}
-MOE_PERMUTE_FUSION=${MOE_PERMUTE_FUSION:-false}
+MOE_PERMUTE_FUSION=${MOE_PERMUTE_FUSION:-true}
+USE_PRECISION_AWARE_OPTIMIZER="${USE_PRECISION_AWARE_OPTIMIZER:-false}"
```
Comment (wenchenvincent) [inline examples/deepseek_v3/train_deepseekv3.sh:181]: Have we tried these options and see which are more performant?

File: examples/deepseek_v3/train_deepseekv3.sh [hunk 3]
```diff
@@ -286,13 +309,21 @@ elif [ $MODEL_SIZE = 16B ]; then
...
     
     moe_options=" \
         --q-lora-rank ${Q_LORA_RANK} \
         --moe-router-num-groups ${EP} \
     "
 fi
+
+# RoPE fusion (Primus-style): enable experimental + fused RoPE when APPLY_ROPE_FUSION is true.
+if [ "${APPLY_ROPE_FUSION:-false}" = true ] || [ "${APPLY_ROPE_FUSION:-0}" = 1 ]; then
+    ROPE_FUSION_OPTS=" --enable-experimental"
```
Comment (wenchenvincent) [inline examples/deepseek_v3/train_deepseekv3.sh:322]: Is the fused rope in TE? Is it the one from AITER?

File: megatron/core/extensions/transformer_engine.py [hunk 4]
```diff
@@ -2735,7 +2735,6 @@ def te_general_gemm(
         return general_gemm(
             A,
             B,
-            workspace=get_workspace(),
```
Comment (wenchenvincent) [inline megatron/core/extensions/transformer_engine.py:2740]: Is this for updating TE version?

File: examples/qwen3/train_qwen3.sh [hunk 5]
```diff
@@ -0,0 +1,769 @@
...
+EVAL_ITERS=${EVAL_ITERS:--1}
+
+GEMM_TUNING="${GEMM_TUNING:-1}"
+USE_GROUPED_GEMM="${USE_GROUPED_GEMM:-true}"
+MOE_USE_LEGACY_GROUPED_GEMM="${MOE_USE_LEGACY_GROUPED_GEMM:-false}"
+MOE_PERMUTE_FUSION="${MOE_PERMUTE_FUSION:-true}"
+NVTE_CK_USES_BWD_V3="${NVTE_CK_USES_BWD_V3:-1}"
+GPT_LAYER_IN_TE="${GPT_LAYER_IN_TE:-true}"
+
+ENABLE_DEEP_EP="${ENABLE_DEEP_EP:-false}"
```
Comment (wenchenvincent) [inline examples/qwen3/train_qwen3.sh:136]: Shall we decide whether to use DEEPEP based on single node or multinode?

File: examples/deepseek_v3/train_deepseekv3.sh
Comment (wenchenvincent) [inline examples/deepseek_v3/train_deepseekv3.sh:121]: Shall we decide whether to use DEEPEP or not based on single node or multi-node?

**Comment by wenchenvincent** [comment]
@sudhu2k  Please also resolve conflicts

File: examples/deepseek_v3/train_deepseekv3.sh (hunk 2 above)
Comment (sudhu2k) [reply examples/deepseek_v3/train_deepseekv3.sh:181]: I've tried MOE_PERMUTE_FUSION option and setting it to true is more performant. I didn't touch the USE_PRECISION_AWARE_OPTIMIZER option since primus was using it for their workload.

File: examples/deepseek_v3/train_deepseekv3.sh (hunk 3 above)
Comment (sudhu2k) [reply examples/deepseek_v3/train_deepseekv3.sh:322]: This is the fused rope from TE.
https://github.com/ROCm/Megatron-LM/blob/e9ad99d6e4dc8917fc0ec0a7dce5684de8473be3/megatron/core/extensions/transformer_engine.py#L2614-L2633

File: megatron/core/extensions/transformer_engine.py (hunk 4 above)
Comment (sudhu2k) [reply megatron/core/extensions/transformer_engine.py:2740]: Yep, From TE 2.10, the api for general_gemm changed where TE removes workspace.
TE 2.8: https://github.com/ROCm/TransformerEngine/blob/a365f2deb555e681bf5b259e344b2bd1fbb4f87c/transformer_engine/pytorch/cpp_extensions/gemm.py#L38

TE 2.10:
https://github.com/ROCm/TransformerEngine/blob/b2a0450804c0a999c42628834290f07d58d671ef/transformer_engine/pytorch/cpp_extensions/gemm.py#L104-L123

File: examples/deepseek_v3/train_deepseekv3.sh (hunk 1 above)
Comment (sudhu2k) [reply examples/deepseek_v3/train_deepseekv3.sh:150]: For DeepSeek-V3, the default is legacy grouped GEMM. Since our CI is using TE 2.10, which has just introduced CK grouped GEMM, it should be fine not to use CK grouped GEMM by default for now. Additionally, during workload testing, we encountered a padding issue even with BF16, which has been fixed by a recent PR: https://github.com/ROCm/TransformerEngine/pull/574.
I could switch the default to use TE’s multistream grouped GEMM, but it would result in a slight performance reduction. I also noticed that for Qwen, the default is TE’s multistream grouped GEMM. I will standardize this by using legacy grouped GEMM as the default for Qwen as well. Let me know what you think!

File: examples/qwen3/train_qwen3.sh (hunk 5 above)
Comment (sudhu2k) [reply examples/qwen3/train_qwen3.sh:136]: I think it’s fine to keep this as an option because the container should have DEEPEP installed and I think most containers people use for Megatron do not include DEEPEP. So, we can allow users to enable it by setting it to True when DEEPEP is available in the container. Let me know what you think!

File: examples/deepseek_v3/train_deepseekv3.sh
Comment (sudhu2k) [reply examples/deepseek_v3/train_deepseekv3.sh:121]: I think it’s fine to keep this as an option because the container should have DEEPEP installed and I think most containers people use for Megatron do not include DEEPEP. So, we can allow users to enable it by setting it to True when DEEPEP is available in the container. Let me know what you think!

File: examples/qwen3/train_qwen3.sh (hunk 5 above)
Comment (sudhu2k) [reply examples/qwen3/train_qwen3.sh:136]: I also have a PR on the mix for enabling DEEPEP and moving ROCm to 7.2
https://github.com/ROCm/Megatron-LM/pull/129
I think we can touch upon this topic after that PR is merged.

**Comment by sudhu2k** [comment]
> @sudhu2k Please also resolve conflicts

Resolved it.

**Review by wenchenvincent — APPROVED** [review]
LGTM. Is the CI failure flaky?
